contracts/cw20-escrow/src/integration_test.rs

```diff
@@ -0,0 +1,141 @@
+        crate::contract::query,
+    );
+    Box::new(contract)
+}
+
+pub fn contract_cw20() -> Box<dyn Contract> {
+    let contract = ContractWrapper::new(
+        cw20_base::contract::handle,
+        cw20_base::contract::init,
+        cw20_base::contract::query,
+    );
+    Box::new(contract)
+}
+
+#[test]
+// receive cw20 tokens and release upon approval
+fn escrow_happy_path_cw20_tokens() {
+    let mut router = mock_router();
+
+    // set personal balance
+    let owner = HumanAddr::from("owner");
+    let init_funds = coins(2000, "btc");
+    router
+        .set_bank_balance(owner.clone(), init_funds.clone())
+        .unwrap();
+
+    // set up cw20 contract with some tokens
+    let cw20_id = router.store_code(contract_cw20());
+    let msg = cw20_base::msg::InitMsg {
+        name: "Cash Money".to_string(),
+        symbol: "CASH".to_string(),
+        decimals: 2,
+        initial_balances: vec![Cw20CoinHuman {
+            address: owner.clone(),
+            amount: Uint128(5000),
+        }],
+        mint: None,
+    };
+    let cash_addr = router
+        .instantiate_contract(cw20_id, &owner, &msg, &[], "CASH")
+        .unwrap();
+
+    // set up reflect contract
+    let escrow_id = router.store_code(contract_escrow());
+    let escrow_addr = router
+        .instantiate_contract(escrow_id, &owner, &InitMsg {}, &[], "Escrow")
+        .unwrap();
+
+    // they are different
+    assert_ne!(cash_addr, escrow_addr);
+
+    // set up cw20 helpers
+    let cash = Cw20Contract(cash_addr.clone());
+
+    // ensure our balances
+    let owner_balance = cash.balance(&router, owner.clone()).unwrap();
+    assert_eq!(owner_balance, Uint128(5000));
+    let escrow_balance = cash.balance(&router, escrow_addr.clone()).unwrap();
+    assert_eq!(escrow_balance, Uint128(0));
+
+    // send some tokens to create an escrow
+    let arb = HumanAddr::from("arbiter");
+    let ben = HumanAddr::from("beneficiary");
+    let id = "demo".to_string();
+    let create_msg = ReceiveMsg::Create(CreateMsg {
+        id: id.clone(),
+        arbiter: arb.clone(),
+        recipient: ben.clone(),
+        end_height: None,
+        end_time: None,
+        cw20_whitelist: None,
+    });
+    let create_bin = to_binary(&create_msg).unwrap();
+    let send_msg = Cw20HandleMsg::Send {
+        contract: escrow_addr.clone(),
+        amount: Uint128(1200),
+        msg: Some(create_bin),
+    };
+    let res = router
+        .execute_contract(&cash_addr, &owner, &send_msg, &[])
+        .unwrap();
+    println!("{:?}", res.attributes);
+    assert_eq!(6, res.attributes.len());
+
+    // ensure balances updated
+    let owner_balance = cash.balance(&router, owner.clone()).unwrap();
+    assert_eq!(owner_balance, Uint128(3800));
+    let escrow_balance = cash.balance(&router, escrow_addr.clone()).unwrap();
+    assert_eq!(escrow_balance, Uint128(1200));
+
+    // ensure escrow properly created
+    let details: DetailsResponse = router
+        .wrap()
```

Comment:
It would be nice it this could be avoided, i.e. handled internally or so. In any case, `get_wrapper()` or `get_querier()` would be a better name? This is exposing something, not wrapping it, AFAIK.

Reply:
`router.as_querier_wrapper()` would be the most explicit. I just felt it was too verbose.

I am happy for better naming, but take a look at some of the big changes in cosmwasm-std for 0.12 when we took in `dyn Querier` and such instead of generics. (And I am happy for better approaches/naming).

Basically, the `Querier trait` only required one method (`raw_query`) and had a lot of default generic implementations. As I made this `&dyn Querier`, we could no longer include generics and defaults on the trait. You can see how [`QuerierWrapper` assumed most of the functionality that used to be in `Querier`](https://github.com/CosmWasm/cosmwasm/blob/master/packages/std/src/traits.rs#L77-L248). And can easily be created from a `Querier`.

I am not sure what a better way to handle this is. I cannot see how to use "deref" magic here to allow `router.query_wasm_smart(...)` directly, which is what I wanted. Please take a look at that wrapper struct (now you understand why) and I am happy for a concrete suggestion to make this cleaner.

Reply:
I see, thanks for the explanation about the need for Wrapper. Naming is tricky... will play with the code base to se if I can propose something concrete.